### PR TITLE
fix(graphql-relational-transformer): ddb references relationships with composite sortkeys

### PR DESCRIPTION
**Pull request by atierian**
#### Description of changes

Adds missing support for references based relationships where the primary model has a primary key with a composite sort key and the Related model uses a DynamoDB table as a data source.

```graphql
type Primary @model {
  tenantId: ID! @primaryKey(sortKeyFields: ["instanceId", "recordId"])
  instanceId: ID!
  recordId: ID!
  content: String
  related: [Related!] @hasMany(references: ["primaryTenantId", "primaryInstanceId", "primaryRecordId"])
}

type Related @model {
  content: String
  primaryTenantId: ID!
  primaryInstanceId: ID!
  primaryRecordId: ID!
  primary: Primary @belongsTo(references: ["primaryTenantId", "primaryInstanceId", "primaryRecordId"])
}
```

In this example, `primaryInstanceId` and `primaryRecordId` are the composite sort key of the global secondary index created on the `Related` model's DynamoDB table. When querying the relationship through the primary model (`Primary.related`), the resolver queries the `Related` table for a GSI where the sort key is `primaryInstanceId#primaryRecordId`. Currently that attribute doesn't exist on the `Related` table, resulting in `related` being null. 

To support composite sort keys, this change adds slots in the mutation pipeline resolvers of the `Related` model. This function adds a new argument to the resolver context prior to the execution of the resolver writing to DynamoDB in the form of key: `primaryInstanceId#primaryRecordId`, value: `<primaryInstanceIdValue>#<primaryRecordIdValue>`.

#### Description of how you validated changes
- Unit tests
- E2E tests: https://us-east-1.console.aws.amazon.com/codesuite/codebuild/594813022831/projects/amplify-category-api-e2e-workflow/batch/amplify-category-api-e2e-workflow:bca6d841-f36f-4859-ba17-d924870c43c0?region=us-east-1
- Manual E2E testing

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-cli/blob/master/CONTRIBUTING.md#tests)
- [ ] ~Relevant documentation is changed or added (and PR referenced)~
- [ ] ~New AWS SDK calls or CloudFormation actions have been added to relevant test and service IAM policies~
- [ ] ~Any CDK or CloudFormation parameter changes are called out explicitly~

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/amplify-graphql-relational-transformer/src/data-source-based-directive-transformer.ts [hunk 1]
```diff
@@ -5,6 +5,7 @@ import {
   TransformerTransformSchemaStepContextProvider,
 } from '@aws-amplify/graphql-transformer-interfaces';
 import { BelongsToDirectiveConfiguration, HasManyDirectiveConfiguration, HasOneDirectiveConfiguration } from './types';
+import { FieldDefinitionNode } from 'graphql';
```
[the remark this answers is not in the document]
Comment (atierian) [reply packages/amplify-graphql-relational-transformer/src/data-source-based-directive-transformer.ts]: Nope, thanks for catching that. Addressed in 9f85b23b0b78de0864371afd7a6fc7660d1c8001

File: packages/amplify-graphql-relational-transformer/src/has-many/has-many-directive-ddb-references-transformer.ts [hunk 2]
```diff
@@ -31,18 +36,28 @@ export class HasManyDirectiveDDBReferencesTransformer implements DataSourceBased
...
     updateTableForReferencesConnection(config, context);
+    updateRelatedModelMutationResolversForCompositeSortKeys(config, context);
     new DDBRelationalReferencesResolverGenerator().makeHasManyGetItemsConnectionWithKeyResolver(config, context);
   };
 
   validate = (context: TransformerContextProvider, config: HasManyDirectiveConfiguration): void => {
+    if (config.indexName) {
+      const mappedObjectName = context.resourceHelper.getModelNameMapping(config.object.name.value);
+      throw new Error(
+        `Invalid @hasMany directive on ${mappedObjectName}.${config.field.name.value} - indexName is not supported with DDB references.`,
```
[the remark this answers is not in the document]
Comment (atierian) [reply packages/amplify-graphql-relational-transformer/src/has-many/has-many-directive-ddb-references-transformer.ts:58]: That's right. I'm coming back to error messages in a separate PR. I've added a link to your comment to make sure it's not missed.

File: packages/amplify-graphql-relational-transformer/src/has-many/has-many-directive-ddb-references-transformer.ts [hunk 3]
```diff
@@ -31,18 +36,28 @@ export class HasManyDirectiveDDBReferencesTransformer implements DataSourceBased
...
-  };
+  transformSchema = (context: TransformerTransformSchemaStepContextProvider, config: HasManyDirectiveConfiguration): void => {};
 
   generateResolvers = (context: TransformerContextProvider, config: HasManyDirectiveConfiguration): void => {
     updateTableForReferencesConnection(config, context);
+    updateRelatedModelMutationResolversForCompositeSortKeys(config, context);
     new DDBRelationalReferencesResolverGenerator().makeHasManyGetItemsConnectionWithKeyResolver(config, context);
   };
 
   validate = (context: TransformerContextProvider, config: HasManyDirectiveConfiguration): void => {
```
[the remark this answers is not in the document]
Comment (atierian) [reply packages/amplify-graphql-relational-transformer/src/has-many/has-many-directive-ddb-references-transformer.ts:54]: Thanks, addressed in 5c2c39f75e21b115cc7b690b4d8d226bec4bd5fb

File: packages/amplify-graphql-relational-transformer/src/resolver/ddb-references-generator.ts [hunk 4]
```diff
@@ -42,6 +43,39 @@ const authFilter = ref('ctx.stash.authFilter');
 const PARTITION_KEY_VALUE = 'partitionKeyValue';
 
 export class DDBRelationalReferencesResolverGenerator extends DDBRelationalResolverGenerator {
+  makeExpression = (references: string[]): ObjectNode => {
+    if (references.length > 1) {
```
[the remark this answers is not in the document]
Comment (atierian) [reply packages/amplify-graphql-relational-transformer/src/resolver/ddb-references-generator.ts:47]: That was a miss. Addressed in 5f0de436fa5a7fd7b656d39d078939c11e419e6f
And added some basic snapshot tests in 0fb215775cae17f336a2ab99daab24cb31f673fe + 99e89ecde76984bf566ed62430d8e9c9347abebe

File: packages/amplify-graphql-relational-transformer/src/resolvers.ts [hunk 5]
```diff
@@ -250,3 +261,181 @@ export const setFieldMappingResolverReference = (
...
+  if (sortKeyFields.length < 2) {
+    return '';
+  }
+
+  return printBlock(`Validate ${keyOperation} mutation for @index '${indexName}'`)(
+    compoundExpression([
+      set(ref(ResourceConstants.SNIPPETS.HasSeenSomeKeyArg), bool(false)),
+      set(ref('keyFieldNames'), list(sortKeyFields.map((f) => str(f)))),
+      forEach(ref('keyFieldName'), ref('keyFieldNames'), [
+        iff(raw('$mergedValues.containsKey("$keyFieldName")'), set(ref(ResourceConstants.SNIPPETS.HasSeenSomeKeyArg), bool(true)), true),
```
[the remark this answers is not in the document]
Comment (atierian) [reply packages/amplify-graphql-relational-transformer/src/resolvers.ts:359]: That's correct.

File: packages/amplify-graphql-relational-transformer/src/resolvers.ts [hunk 6]
```diff
@@ -250,3 +261,181 @@ export const setFieldMappingResolverReference = (
...
+
+  const objectName = ctx.output.getMutationTypeName();
+  if (!objectName) {
+    throw new Error(`
+      Mutation type name is undefined when updated mutation resolvers for composite sortKeys used in DDB references based relationships.
+      This should not happen, please file a bug at https://github.com/aws-amplify/amplify-category-api/issues/new/choose`);
+  }
+
+  const capitalizedName = toUpper(relatedType.name.value);
+  const createResolver = ctx.resolvers.getResolver(objectName, `create${capitalizedName}`);
```
[the remark this answers is not in the document]
Comment (atierian) [reply packages/amplify-graphql-relational-transformer/src/resolvers.ts]: > Why would we need to add this slot to a delete resolver?

We don't. The delete mutation slot was incorrectly included in that commit. Addressed in edd9eee97848f54ef394576946ef7abd444602dc

> Update makes sense IFF the reference fields are changing.

We create the concatenated composite key value if the composite fields are included.
The logic is:
```mermaid
flowchart TD
 A[Entry Point] --> B{Any sort key value included?};
 B -- Yes --> C{All sort key values included?};
 B -- No --> D[Nothing to do];
 C -- Yes --> E[put concatenated sort key to $ctx.args.input];
 C -- No --> F[Error: When creating any part of the composite sort key for global secondary index <index-name> you must provide all fields for the key. Missing key: <key-field-name>];
```

**Review by palpatim — APPROVED** [review]

**Review by palpatim — APPROVED** [review]

**Review by phani-srikar — APPROVED** [review]

**Pull request merged**